File: editor/src/dsp/dynamics.ts
import { DYN_KNEE_DB, type Band } from './bands'

export interface DynStep {
  /** Smoothed engagement, 0..1. */
  env: number
  /** Gain offset to apply to the band, in dB. */
  delta: number
}

/**
 * Engagement across the soft knee, 0..1, from how far the level sits past the
 * threshold in the engaging direction.
 *
 * Smoothstep rather than a straight ramp. A ramp is continuous but its slope
 * jumps at both corners, so the band goes from perfectly still to travelling at
 * full rate the instant the level touches the threshold — the grabbiness a soft
 * knee exists to remove. Both agree at half a knee in, where engagement is 0.5.
 *
 * Mirrors `knee` in `src/dsp/dynamics.rs`.
 */
export function knee(overDb: number): number {
  const t = Math.min(Math.max(overDb / DYN_KNEE_DB, 0), 1)
  return t * t * (3 - 2 * t)
}

/** One-pole move from `env` toward `target` over `tauMs`, `dt` seconds on. */
export function stepToward(env: number, target: number, tauMs: number, dt: number): number {
  const tau = Math.max(tauMs / 1000, 0.001)
  const next = env + (target - env) * (1 - Math.exp(-dt / tau))
  // A tail this far down is orders below anything a gain could show.
  return Math.abs(next) < 1e-6 ? 0 : next
}

/**
 * One control-rate step of a band's dynamics.
 *
 * The level's distance past the threshold maps across a `DYN_KNEE_DB` soft knee
 * to 0..1, which is then smoothed by a one-pole using attack going up and
 * release coming down. The band's gain moves by `env * dynRange`.
 */
export function dynamicStep(
  band: Pick<Band, 'threshold' | 'dynMode' | 'dynRange' | 'attack' | 'release'>,
  levelDb: number,
  env: number,
  dt: number,
): DynStep {
  const over = band.dynMode === 'above' ? levelDb - band.threshold : band.threshold - levelDb
  const target = knee(over)

  const tauMs = target > env ? band.attack : band.release
  const next = stepToward(env, target, tauMs, dt)

  return { env: next, delta: next * band.dynRange }
}

/** RMS of a time-domain block, in dBFS. */
export function rmsDb(buf: Float32Array): number {
  let sum = 0
  for (let i = 0; i < buf.length; i++) sum += buf[i] * buf[i]
  return 20 * Math.log10(Math.max(Math.sqrt(sum / buf.length), 1e-6))
}

File: src/dsp/dynamics.rs
//! Band dynamics — the envelope that moves a band's gain with the signal.
//!
//! Ported from `editor/src/dsp/dynamics.ts` so the UI's prediction of where a
//! dynamic band sits matches what the audio thread actually does.

use crate::params::{DynMode, DYN_KNEE_DB};

/// A band's dynamics settings, as they stand for one control block.
#[derive(Clone, Copy, Debug)]
pub struct DynSettings {
    pub threshold_db: f32,
    pub mode: DynMode,
    /// Signed dB the band travels at full engagement.
    pub range_db: f32,
    pub attack_ms: f32,
    pub release_ms: f32,
}

/// One control-rate step.
///
/// `over` is how far the measured band level sits past the threshold in the
/// engaging direction; it maps across a [`DYN_KNEE_DB`] soft knee to 0..1, then
/// is smoothed by a one-pole using attack going up and release coming down.
/// Returns the new envelope and the gain offset in dB.
pub fn dynamic_step(s: DynSettings, level_db: f32, env: f32, dt: f32) -> (f32, f32) {
    let over = match s.mode {
        DynMode::Above => level_db - s.threshold_db,
        DynMode::Below => s.threshold_db - level_db,
    };
    let target = (over / DYN_KNEE_DB).clamp(0.0, 1.0);

    let tau_ms = if target > env {
        s.attack_ms
    } else {
        s.release_ms
    };
    let tau = (tau_ms / 1000.0).max(0.001);
    let next = env + (target - env) * (1.0 - (-dt / tau).exp());

    (next, next * s.range_db)
}

/// Mean square of a block, as dBFS. The floor keeps silence from reaching -inf.
pub fn ms_to_db(mean_square: f32) -> f32 {
    10.0 * mean_square.max(1e-12).log10()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn settings(
        threshold_db: f32,
        mode: DynMode,
        range_db: f32,
        attack_ms: f32,
        release_ms: f32,
    ) -> DynSettings {
        DynSettings {
            threshold_db,
            mode,
            range_db,
            attack_ms,
            release_ms,
        }
    }

    #[test]
    fn silence_leaves_the_band_alone() {
        let (env, delta) = dynamic_step(
            settings(-24.0, DynMode::Above, -6.0, 20.0, 200.0),
            -90.0,
            0.0,
            0.01,
        );
        assert_eq!(env, 0.0);
        assert_eq!(delta, 0.0);
    }

    #[test]
    fn a_loud_band_walks_toward_full_range() {
        let mut env = 0.0;
        let mut delta = 0.0;
        // Well past the knee, so the target is 1.0 the whole way.
        for _ in 0..200 {
            let step = dynamic_step(
                settings(-24.0, DynMode::Above, -6.0, 20.0, 200.0),
                -6.0,
                env,
                0.001,
            );
            env = step.0;
            delta = step.1;
        }
        assert!(env > 0.99, "env was {env}");
        assert!((delta + 6.0).abs() < 0.1, "delta was {delta}");
    }

    #[test]
    fn below_mode_engages_on_quiet_signal() {
        let (env, delta) = dynamic_step(
            settings(-24.0, DynMode::Below, 4.0, 1.0, 200.0),
            -40.0,
            0.0,
            1.0,
        );
        assert!(env > 0.9);
        assert!(delta > 3.0);
    }

    #[test]
    fn the_knee_gives_partial_engagement() {
        // Half a knee past the threshold should target 0.5, not 1.0.
        let (env, _) = dynamic_step(
            settings(-24.0, DynMode::Above, -6.0, 0.1, 0.1),
            -24.0 + DYN_KNEE_DB / 2.0,
            0.0,
            1.0,
        );
        assert!((env - 0.5).abs() < 0.01, "env was {env}");
    }

    #[test]
    fn release_is_slower_than_attack() {
        let (fast, _) = dynamic_step(
            settings(-60.0, DynMode::Above, -6.0, 10.0, 500.0),
            0.0,
            0.0,
            0.005,
        );
        let (slow, _) = dynamic_step(
            settings(-60.0, DynMode::Above, -6.0, 10.0, 500.0),
            -90.0,
            1.0,
            0.005,
        );
        assert!(fast > 0.3, "attack moved only to {fast}");
        assert!(slow > 0.98, "release moved all the way to {slow}");
    }
}
